### PR TITLE
Add exec output

### PR DESCRIPTION
**Pull request by amitaekbote**
Add common shelling out function to exec.

it returns stdout, stderr and error

### CODE REVIEW

**Comment by amitaekbote** [comment]
@darkonie do you think we need to return the exit status or stderr would do? since this is a generic function you can combine the buf's for the runner

**Comment by amitaekbote** [comment]
@collinvandyck @darkonie ready for review

File: dcos/http/transport/roundtripper.go [hunk 1]
```diff
@@ -94,10 +94,10 @@ func (t *dcosRoundtripper) GenerateToken() error {
...
 	defer t.Unlock()
 
-	// TODO: this is very broken with the latest `jwt-go` lib version 3.0.0
-	token := jwt.New(jwt.SigningMethodRS256)
-	token.Claims["uid"] = t.uid
-	token.Claims["exp"] = time.Now().Add(t.expire).Unix()
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"uid": t.uid,
+		"exp": time.Now().Add(t.expire).Unix(),
+	})
```
Comment (darkonie) [inline dcos/http/transport/roundtripper.go]: this does not look related to this PR, can we please undo it?

File: exec/exec.go [hunk 2]
```diff
@@ -79,3 +82,73 @@ func Run(ctx context.Context, command string, arg []string) (*CommandExecutor, e
...
+
+	var (
+		// define an empty cancel function
+		cancel context.CancelFunc = func() {}
+		arg    []string
+	)
+
+	if ctx == nil {
+		// default to 10 seconds timeout.
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second*10)
```
Comment (darkonie) [inline exec/exec.go]: let's extract the 10 sec constant

File: exec/exec.go [hunk 3]
```diff
@@ -79,3 +82,73 @@ func Run(ctx context.Context, command string, arg []string) (*CommandExecutor, e
 
 	return commandExecutor, nil
 }
+
+// Output returns stdout, stderr and error status for a given shell command
+func Output(ctx context.Context, command []string, timeout ...string) ([]byte, []byte, error) {
```
Comment (darkonie) [inline exec/exec.go]: why do you have a list of timeouts?

File: exec/exec.go [hunk 4]
```diff
@@ -79,3 +82,73 @@ func Run(ctx context.Context, command string, arg []string) (*CommandExecutor, e
...
+	if ctx == nil {
+		// default to 10 seconds timeout.
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second*10)
+	}
+
+	if len(command) == 0 {
+		return nil, nil, errors.New("unable to execute a command with empty Cmd field")
+	}
+
+	if timeout != nil {
```
Comment (darkonie) [inline exec/exec.go]: this is not the right way to compare strings in go
you should do
`if timeout != "" {`
or
`if len(timeout) == 0 {`

File: exec/exec.go (hunk 3 above)
Comment (darkonie) [inline exec/exec.go]: can we change the signature to take a `command ...string`? it looks nicer then array of strings

File: exec/exec.go [hunk 5]
```diff
@@ -79,3 +82,73 @@ func Run(ctx context.Context, command string, arg []string) (*CommandExecutor, e
...
+		timeval, err := time.ParseDuration(timeout[0])
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "invalid timeout string %s", timeout)
```
Comment (darkonie) [inline exec/exec.go]: did the unittest pass? this errors is not in vendor folder, you need another commit to update vendor folder with https://github.com/pkg/errors

File: scripts/test.sh [hunk 6]
```diff
@@ -33,6 +33,7 @@ function _gofmt {
 
 function _goimports {
     logmsg "Running 'goimports' ..."
+    go get github.com/pkg/errors
```
Comment (darkonie) [inline scripts/test.sh]: this does not belong here, you should update vendor folder

File: scripts/test.sh [hunk 7]
```diff
@@ -96,7 +97,7 @@ function main {
     # Dependencies required for testing
     go get -u github.com/kardianos/govendor
     govendor init
-    govendor fetch github.com/dgrijalva/jwt-go/^::gopkg.in/dgrijalva/jwt-go.v2
+    govendor fetch github.com/dgrijalva/jwt-go/^::gopkg.in/dgrijalva/jwt-go.v3
```
Comment (darkonie) [inline scripts/test.sh]: this is not related to this PR, and also did you validate that this works? last time i tried it it did not work for some reason...

File: exec/exec.go (hunk 4 above)
Comment (collinvandyck) [inline exec/exec.go]: timeout in this case is a slice of string `timeout... string`.   that being said, it's confusing why timeout is a variadic parameter but later on you're only using `timeout[0]`.   ideally you'd pass a timeout as a time.Duration into the function, not a string.

Comment (amitaekbote) [inline exec/exec.go]: Ill change this, I was playing with golang here, will just pass the time.Duration to a function

File: exec/exec.go (hunk 3 above)
Comment (amitaekbote) [inline exec/exec.go]: sure

File: exec/exec.go (hunk 5 above)
Comment (amitaekbote) [inline exec/exec.go]: I added it manually in my workspace, will add to vendor dir and to the pr

File: exec/exec_test.go [hunk 8]
```diff
@@ -86,3 +87,12 @@ func TestBadReturnCode(t *testing.T) {
...
 }
+
+func TestOutput(t *testing.T) {
+
+	out, stderr, err := Output(nil, 0, "echo", "hello")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Printf("output %s %s \n", out, stderr)
+}
```
Comment (darkonie) [inline exec/exec_test.go]: could we run a more command that write to both stdout and stderr and then validates the result instead of just checking for error?

File: exec/exec_test.go [hunk 9]
```diff
@@ -86,3 +87,37 @@ func TestBadReturnCode(t *testing.T) {
...
+func TestOutput(t *testing.T) {
+
+	out, stderr, err := Output(nil, "echo", "hello")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Printf("output %s %s \n", out, stderr)
+
+	// Expect error
+	output, stdaerr, errtrue := Output(nil, "ec", "hello")
```
Comment (darkonie) [inline exec/exec_test.go]: type, `stderr`

File: exec/exec_test.go [hunk 10]
```diff
@@ -86,3 +87,37 @@ func TestBadReturnCode(t *testing.T) {
...
+	_, _, err := Output(ctx, "sleep", "4")
+	if err == nil {
+		t.Fatal("expect error got nil")
+	}
+}
+
+func TestOutputTimeoutPass(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	_, _, err := Output(ctx, "sleep", "4")
```
Comment (darkonie) [inline exec/exec_test.go]: maybe lets move to milliseconds? 4 seconds is a lot
